From a87eba2d404c1d30d69c9968427053598e8ebb31 Mon Sep 17 00:00:00 2001
From: "Zhou, Lihua" <lihuax.zhou@intel.com>
Date: Fri, 9 Feb 2018 09:59:55 +0800
Subject: [PATCH 16/16] Add recovery message into BCB to support adb reboot
 recovery

Write bootone-bootloader, bootonce-fastboot or bootonce-recovery
into misc partition according to reboot target.
It needs corresponding function write_reboot_bootloader() defined
in bootable/recovery/bootloader_message/bootloader_message.cpp.

Change-Id: Ic25adf5facab89caae431fa612d4e393d7d6638c
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-56251
Signed-off-by: Zhou, Lihua <lihuax.zhou@intel.com>
---
 init/reboot.cpp | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/init/reboot.cpp b/init/reboot.cpp
index 5c7ddf1..f296314 100644
--- a/init/reboot.cpp
+++ b/init/reboot.cpp
@@ -483,15 +483,31 @@ bool HandlePowerctlMessage(const std::string& command) {
         cmd = ANDROID_RB_RESTART2;
         if (cmd_params.size() >= 2) {
             reboot_target = cmd_params[1];
-            // When rebooting to the bootloader notify the bootloader writing
+            // When rebooting to the bootloader, notify the bootloader writing
             // also the BCB.
+            std::string err;
+            std::string cmd;
             if (reboot_target == "bootloader") {
-                std::string err;
-                if (!write_reboot_bootloader(&err)) {
+                cmd = "bootonce-bootloader";
+                if (!write_reboot_bootloader(&err, &cmd)) {
                     LOG(ERROR) << "reboot-bootloader: Error writing "
                                   "bootloader_message: "
                                << err;
                 }
+            } else if (reboot_target == "fastboot") {
+                cmd = "bootonce-fastboot";
+                if (!write_reboot_bootloader(&err, &cmd)) {
+                    LOG(ERROR) << "reboot-fastboot: Error writing "
+                                  "bootloader_message: "
+                               << err;
+                }
+            } else if (reboot_target == "recovery") {
+                cmd = "bootonce-recovery";
+                if (!write_reboot_bootloader(&err, &cmd)) {
+                    LOG(ERROR) << "reboot-recovery: Error writing "
+                                  "bootloader_message: "
+                               << err;
+                }
             }
             // If there is an additional bootloader parameter, pass it along
             if (cmd_params.size() == 3) {
-- 
1.9.1

